### ISSUE TITLE
Accessibility: Comply with WCAG 2.2

### ISSUE THREAD

**Issue by personalizedrefrigerator** [laurent22/joplin#10795]
# Summary

\[ [Progress: Desktop](#desktop) \] | \[ [Progress: Mobile](https://github.com/laurent22/joplin/issues/11661) \]

This issue tracks Joplin's progress towards compliance with the [WCAG (Web Content Accessibility Guidelines) v2.2](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.2#navigable).

| Level | Total | Done | % Done |
|-------|-------|------|--------|
| A     | 75 | 60 | 80 |
| AA    | 40 | 18 | 45 |
| Total | 115 | 78 | 67.8 |

Updated at 2025-02-06

## About the WCAG

The [WCAG (Web Content Accessibility Guidelines)](https://www.w3.org/TR/WCAG22/) is a set of guidelines for improving web accessibility.

The WCAG is divided into success criteria. Each is labelled with "Level A", "Level AA", or "Level AAA". These correspond to different [conformance levels](https://www.w3.org/WAI/WCAG22/Understanding/conformance#conf-req1).

## Conformance levels

There are three different conformance levels[^1]:

- Level A: All Level A success criteria are satisfied.
    - This is the lowest level of conformance.
- Level AA: All Level A **and** Level AA success criteria are satisfied.
- Level AAA: **All** success criteria are satisfied.

According to WebAIM.org[^2],

> Most websites should aim for Level AA conformance. Depending on the target users, some Level AAA success criteria such as [Reading Level,](https://webaim.org/standards/wcag/checklist#sc3.1.5) [Contrast (Enhanced),](https://webaim.org/standards/wcag/checklist#sc1.4.6) [Target Size](https://webaim.org/standards/wcag/checklist#sc2.5.5), or others may be beneficial to implement. Very few websites can claim full Level AAA conformance—not even webaim.org or the WCAG website itself!

The WCAG's definition of the different conformance levels[^1] confirms WebAIM's statement that Level AAA is not always achievable:

> [!NOTE]
> 
> It is not recommended that Level AAA conformance be required as a general policy for entire sites because it is not possible to satisfy all Level AAA success criteria for some content.

Based on the above, for now, this issue targets **Level AA** conformance with the WCAG.

[^1]: https://www.w3.org/WAI/WCAG22/Understanding/conformance#conf-req1
[^2]: https://webaim.org/standards/wcag/#current

## This issue: General tasks

- [ ] Find existing accessibility issues & link them to this task.
- [ ] Go through the [WCAG 2.2 quick reference](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.2&showtechniques=241%2C243#keyboard-accessible) and evaluate each task. Create issues for failures.

**Notes**: In this issue, different task priorities are marked differently. For example,

- [ ] [❗] This is a task that **must** be completed.
	- This should be used mostly for level A tasks.
- [ ] [🟡] This is a task that *should* be completed.
	- This should be used mostly for level AA tasks.
- [ ] \[🔷\] This is a low priority task that *may* be completed.
- [ ] \[🔎\] The priority of this task is unclear and requires additional research.

> [!NOTE]  
> The tasks below are a work-in-progress and **based on** the WCAG 2.2 spec. Completing the tasks below does not guarantee compliance with the spec.

# Quick list of general issues

- **Mobile**:
	- [x] Accessibility focus visits nonfunctional "new note" and "new to-do" buttons.
		- Fixed for now by changing the component used for the new note menu.
		- Upstream issue will be fixed by https://github.com/callstack/react-native-paper/pull/4498, but it's taking a long time for the pull request to get reviewed. For now, it might make sense to patch `react-native-paper`.
	- [ ] Settings tabs might be missing `role` information (not announced as buttons). ([SC 4.1.2](https://www.w3.org/WAI/WCAG22/quickref/#name-role-value))
	- [ ] Search result count announcements. ([SC 4.1.3](https://www.w3.org/WAI/WCAG22/Understanding/status-messages.html))
- **Desktop:**
	- [x] #11640
	- [x] #11639
	- [x] Scrollbars are too thin ([SC 2.5.8](https://www.w3.org/WAI/WCAG22/quickref/#target-size-minimum)) -- https://github.com/laurent22/joplin/pull/11627
	- [x] Scrollbars have too low contrast in some cases ([SC 1.4.3](https://www.w3.org/WAI/WCAG22/quickref/#contrast-minimum)) -- https://github.com/laurent22/joplin/pull/11708
	- [x] Font picker suggestions aren't accessible
	    - See https://github.com/laurent22/joplin/pull/11763
	- [ ] <kbd>ctrl</kbd>-<kbd>m</kbd> keyboard shortcut for toggling tab focus/indentation is not documented.
	- [x] #11621
	- [ ] (Maybe?) Command-related changes to the editor are not announced. ([SC 4.1.3](https://www.w3.org/WAI/WCAG22/Understanding/status-messages.html))
		- We discussed this previously and the pull request was closed due to added complexity.

# Desktop
<div id="desktop"></div>

> [!NOTE]
>
> Several relevant changes were made shortly before this issue was created:
> - https://github.com/laurent22/joplin/issues/9982
> - https://github.com/laurent22/joplin/commit/b108bf799d1f90fcee2069371278f311d3913568
> - https://github.com/laurent22/joplin/commit/dd5240d018e0a3ffc101f922b61c6e3498b99816
> - https://github.com/laurent22/joplin/commit/4827d0bf9278ad6e133977558e6751b3659af2e0

## [1\. Perceivable](https://www.w3.org/TR/WCAG20/#perceivable)

**Summary:** All non-decorative components of the user interface should be described in a way perceivable to the user.

### 1.1.1 [Non-text content](https://www.w3.org/WAI/WCAG22/quickref/#non-text-content) (Level A)

> All non-text content that is presented to the user has a text alternative that serves the equivalent purpose, except for situations listed below.

To meet this requirement, all inputs controls and inputs must be labeled.

Tasks:

- [x] Find and fix unlabeled inputs and buttons in Joplin.
    - [x] [❗] "Edit", "Save", and "Copy" buttons in the note properties dialog -- b108bf799d1f90fcee2069371278f311d3913568
    - [x] [❗] Whether toolbar buttons are disabled or not -- b108bf799d1f90fcee2069371278f311d3913568
    - [x] [❗] Mermaid chart download button.
    - [x] [❗] GotoAnything help button -- https://github.com/laurent22/joplin/pull/10816.
    - [x] [❗] Note attachments screen: Reverse sort order buttons missing labels -- https://github.com/laurent22/joplin/pull/11231.
    - [x] [❗] Master password dialog: Show password buttons -- https://github.com/laurent22/joplin/pull/11231.
    - [x] [❗] Master password dialog: Inputs unlabelled -- https://github.com/laurent22/joplin/pull/11231.
    - [x] [❗] Master password dialog: Valid icon missing label -- https://github.com/laurent22/joplin/pull/11231.
    - [x] [❗] Edit button for each item on the "Keyboard shortcuts" page -- https://github.com/laurent22/joplin/pull/11232.
    - [x] [❗] Rich Text Editor formatting button -- https://github.com/laurent22/joplin/pull/11655
    - [x]  [❗] Buttons in "Change application layout" screen -- https://github.com/laurent22/joplin/pull/11718.
    - [x] \[🔎\] Search for more issues (focus on the note lists & settings).
- [x] [🟡] Find and fix improperly labeled inputs (where the label is improperly associated with the input).
    - [x] See https://www.w3.org/WAI/WCAG22/Techniques/html/H44
- [x] [❗] Hide **decoration-only** icons that are currently read as characters.
    - [x] Hide the "symbol e908" read before "all notes" and the notebook icon just before "NOTEBOOKS" in the sidebar -- https://github.com/laurent22/joplin/pull/11196
    - [x] Hide the decoration-only cloud provider icons in the sync wizard -- https://github.com/laurent22/joplin/pull/11649.
- [x] [❗]  #11642
    - [x] Are there decorative images in the welcome notes that should be hidden?
    - [x] Are there non-decorative images that need descriptions?
- [ ] [❗] Rich Text Editor code blocks: 
    -  \[🔎\] May depend on the screen reader.

### [1.2.1. Audio-only and Video-only](https://www.w3.org/TR/UNDERSTANDING-WCAG20/media-equiv.html), [1.2.2. Captions (Prerecorded)](https://www.w3.org/WAI/WCAG20/quickref/#qr-media-equiv-captions), and [1.2.3. Audio Description or Media Alternative (Prerecorded)](https://www.w3.org/WAI/WCAG20/quickref/#qr-media-equiv-audio-desc)

Joplin allows users to include prerecorded audio and video in notes. When included as a link (e.g. `[link](:/resourceidhere000000000000000000)`), the generated HTML video element has no associated description.

In the **desktop app**, we should:

- [x] \[🔎\] How can we make it easy for users to add alternative text to `image`s, `video`s and `audio` elements?
    - I think this is mostly a problem in the Rich Text Editor. In the RTE, I'm not aware of a way to add ALT text.

Also see [the spec](https://www.w3.org/TR/WCAG20/#perceivable) for related tasks.

### [1.3 Adaptable](https://www.w3.org/TR/WCAG20/#content-structure-separation)

**Summary**: Content must be presentable in different ways without losing information.

### 1.3.1. Info and Relationships (Level A)

> Information, structure, and relationships conveyed through presentation can be programmatically determined or are available in text. (Level A)

- [x] [❗] [Use ARIA landmarks](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA11.html) for important regions.
    - [x] [❗] #11644
    - [x] [❗] Button bars on secondary screens should have `role=navigation` -- https://github.com/laurent22/joplin/pull/11650
    - [x] \[🔎\] Wrap the editor with `role=main` -- https://github.com/laurent22/joplin/pull/11246
    - [x] [❗] Wrap navigation (the sidebars?) with `role=navigation` -- https://github.com/laurent22/joplin/pull/11246
        - [x] Related: [Guidance for multiple landmarks of the same type: Give them different labels](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA11.html#example-2-multiple-landmarks-of-the-same-type-and-aria-labelledby).
- [x] [❗] [Use `aria-labelledby` (or perhaps `aria-label`) for groups and landmarks](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA13.html)
    - [x] Group & associate descriptions with different settings controls -- https://github.com/laurent22/joplin/pull/10812.
- [x] [Use `label` elements to label form controls](https://www.w3.org/WAI/WCAG22/Techniques/html/H44.html)
    - [x] [❗] Settings text & number inputs -- https://github.com/laurent22/joplin/pull/10812
    - [x] [❗] Settings spinners -- https://github.com/laurent22/joplin/pull/10812
    - [x] [❗] Already done for checkbox inputs.
    - [x] [❗] Note properties dialog: Inputs -- https://github.com/laurent22/joplin/pull/11700
    - [x] [❗] Font picker: Suggestions list structure/connection to input -- https://github.com/laurent22/joplin/pull/11763
- [x] [Mark icon font icons as icons](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA24.html).  
    **Note**: This is important even if the icons are hidden from a screen reader — some accessibility tools replace fonts on a page for readability. Marking icons created from a font as such should prevent these accessibility tools from treating them as text.
    - [x] [❗]   Settings sidebar -- https://github.com/laurent22/joplin/pull/10812
    - [x] [❗]  Settings "toggle advanced" -- https://github.com/laurent22/joplin/pull/10812
    - [x] [❗]  Expand/contract notebook icon
    - [x] [❗]  Command icons (markdown editor toolbar icons, some Rich Text editor toolbar icons)
    - [x] [❗]  Check icons in the sync wizard -- https://github.com/laurent22/joplin/pull/11649
    - [x] [❗]  Trash folder icon
- [ ] [❗] Use some method other than whitespace characters or padding to mark something as a subgroup of another. [Related to F33](https://www.w3.org/WAI/WCAG22/Techniques/failures/F33.html).  
    **Effect**: It's currently impossible or difficult to tell whether some item is a subsection of another.
    - [ ] Settings sidebar: Settings subsections.
    - [x] Sidebar: Subnotebooks: Consider using [`aria-level` along with the `tree` pattern](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) -- https://github.com/laurent22/joplin/pull/11196
- [ ] [❗] #11637
- [x] [❗] Sync wizard: Lists and headings aren't identified as such -- https://github.com/laurent22/joplin/pull/11649
- [x] \[🔎\] Does Joplin use `::before` or `::after` to include non-decorative content (see [F87](https://www.w3.org/WAI/WCAG22/Techniques/failures/F87.html))?
    - Used in the RTE for lists. This may be difficult to improve.
- [x] \[🔎\] Does Joplin use structural markup to achieve a visual-only effect? ([F43](https://www.w3.org/WAI/WCAG22/Techniques/failures/F43.html) and related).
- [x] [❗]  Warning messages shown at the top of the screen -- https://github.com/laurent22/joplin/pull/11775

### 1.3.2. Meaningful Sequence (Level A)

> When the sequence in which content is presented affects its meaning, a correct reading sequence can be programmatically determined. (Level A)

- [x] Joplin currently adjusts focus order by overriding <kbd>tab</kbd> and <kbd>shift</kbd>\-<kbd>tab</kbd>. See [SC 2.4.3](https://www.w3.org/WAI/WCAG22/Understanding/focus-order).
- [x] \[🔎\] Meaningful sequences can be particularly easy to disrupt using `flex-order` or similar order-changing CSS. Are we using `flex-order` anywhere? See [F1](https://www.w3.org/WAI/WCAG22/Techniques/failures/F1.html).
- [x] \[🔎\] Does this include RTL support? ([Related issue](https://github.com/laurent22/joplin/issues/3991)).

### 1.3.3. [Sensory Characteristics](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-understanding.html) (Level A)

> Instructions provided for understanding and operating content do not rely solely on sensory characteristics of components such as shape, size, visual location, orientation, or sound. (Level A)

**Note**: §1.4 covers color guidelines.

This task refers to **instructions**.  
For example, "press the red button in the bottom right corner of the screen to edit" would be bad.

We should:

- [x] [❗] #11646
- [x] [❗] #11647
- [x] (Different task) Check the documentation website.

### 1.3.4. [Orientation](https://www.w3.org/WAI/WCAG22/Understanding/orientation.html) (Level AA)

> Content does not restrict its view and operation to a single display orientation, such as portrait or landscape, unless a specific display orientation is [essential](https://www.w3.org/WAI/WCAG22/Understanding/orientation.html#dfn-essential).

Currently, Joplin doesn't restrict devices to a single orientation. However, it may be difficult to use the desktop app on a phone-sized device in portrait mode. In particular, without moving the sidebars from "View" > "Change application layout".

- [x] \[🔎\] Does this require better by-default support for small screens? If so, a possible solution could be to minimize secondary panels (e.g. by moving them to an overflow tab) when low on space. *(Note: I do not think we need to support this, simply because we are not a website but an app designed for a specific platform. The desktop app is for relatively wide screens, and for smaller screens we have the mobile app)*
    - [x] This might be part of [§1.4.10](https://www.w3.org/WAI/WCAG22/Understanding/reflow.html).

## 1.3.5. [Identify input purpose](https://www.w3.org/WAI/WCAG22/quickref/#identify-input-purpose) (AA)

> The purpose of each input field collecting information about the user can be programmatically determined when:
> - The input field serves a purpose identified in the Input Purposes for User Interface Components section; and
> - The content is implemented using technologies with support for identifying the expected meaning for form input data.

In particular
> The intent of this Success Criterion is to ensure that the purpose of a form input collecting information about the user can be programmatically determined, so that user agents can extract and present this purpose to users using different modalities.
>
> — https://www.w3.org/WAI/WCAG22/Understanding/identify-input-purpose.html

See [input-purposes](https://www.w3.org/TR/WCAG21/#input-purposes) for a list of input purposes such that
> When these user input purposes are present, and if the technology supports doing so, the field purpose must be programmatically identifiable.

- [x] [🟡] Switch to a date picker that doesn't use `type=text` for setting note alarms (perhaps use `<input type=datetime-local>`). https://github.com/laurent22/joplin/pull/11725
- [x] [🟡] Switch to `type=search` for search inputs
    - Partially resolved by https://github.com/laurent22/joplin/pull/11645. There may be other search inputs.
- [ ] Set `autocomplete=` attributes in settings
    - [ ] For the username of the "Nextcloud" and "WebDav" sync target - set it to "username"
    - [ ] For the email of the Joplin Server sync target - set it to "email"

### 1.4 [Distinguishable](https://www.w3.org/TR/WCAG22/#distinguishable)

> Make it easier for users to see and hear content including separating foreground from background

### [Color for distinguishing content](https://www.w3.org/TR/WCAG22/#use-of-color) (§1.4.1, A):

- [ ] \[🔎\] Check color contrast between neighboring content.
    - [ ] Add a high-contrast theme for the markdown editor?

### [Audio control](https://www.w3.org/TR/WCAG22/#audio-control) (§1.4.2, A)

> If any audio on a Web page plays automatically for more than 3 seconds, either a mechanism is available to pause or stop the audio, or a mechanism is available to control audio volume independently from the overall system volume level. 

At present, all audio should be user-created (and only auto-play if this is set by the user in HTML). Additionally, audio included with `[](:/resourceidhere)` has Electron-provided controls.

### [Minimum contrast](https://www.w3.org/TR/WCAG22/#contrast-minimum) (§1.4.3, AA):

- [x] [🟡] Improve color contrast of "Conflicts" in the sidebar with the default theme. (Currently uses 3.32:1 contrast) -- https://github.com/laurent22/joplin/pull/11638
- [x] [🟡] Improve the color contrast of "All notes" in the sidebar. (Currently has 3.35:1 contrast) --- https://github.com/laurent22/joplin/pull/11638
	- When active, "All notes" has sufficient contrast.
- [x] [🟡] Improve color contrast of setting item descriptions -- https://github.com/laurent22/joplin/pull/11246
- [x] [🟡] Improve link URL contrast (currently has 3.81:1 contrast) -- https://github.com/laurent22/joplin/pull/11635
- [x] [🟡] Improve notebook list scrollbar contrast in the default light theme -- https://github.com/laurent22/joplin/pull/11708
	- Current contrast is [1.22:1](https://webaim.org/resources/contrastchecker/?fcolor=40434A&bcolor=313640) when not active, [1.6:1](https://webaim.org/resources/contrastchecker/?fcolor=535457&bcolor=313640) when active.
- [ ] [🟡] Improve code block contrast in the default light theme
    - For example, the current [contrast ratio for some highlighted text is 3.87:1](https://webaim.org/resources/contrastchecker/): ![Screenshot: Dev tools indicates that the contrast of the function name "test" in a JavaScript code block has low contrast](https://github.com/user-attachments/assets/478921c6-7ede-461c-8f8e-fd8d913eaa1c)

**Note**: It might not be necessary to improve contrast of inactive buttons (despite having low contrast). See [discussion](https://ux.stackexchange.com/a/80885) and "incidental" in §1.4.3.

### [Resize text](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.2&showtechniques=131%2C132%2C241%2C243%2C134#resize-text) (§1.4.4, AA):

Joplin already supports zoom with <kbd>ctrl</kbd>\-<kbd>+</kbd> and <kbd>ctrl</kbd>\-<kbd>\-</kbd>

- [x] [🟡] #11654

**Avoid images of text unless customizable/essential (§1.4.5, AA)**:

- [x] For the most part, Joplin doesn't seem to rely on images of text. However, the welcome notes does include `SubNotebooks.png`, which includes text that isn't presented elsewhere. Addressed in https://github.com/laurent22/joplin/pull/11643

**Note**: See the WCAG spec for §1.4.6-1.4.9 (all level AAA).

### 1.4.10: [Reflow](https://www.w3.org/WAI/WCAG22/quickref/#reflow) (AA)

> Content can be presented without loss of information or functionality, and without requiring scrolling in two dimensions for:
> 
> - Vertical scrolling content at a width equivalent to 320 CSS pixels;
> - Horizontal scrolling content at a height equivalent to 256 CSS pixels;
> 
> Except for parts of the content which require two-dimensional layout for usage or meaning.

- [ ] [🟡] By default, a width of 320 CSS pixels shows only the notebooks list and part of the note list (everything else is hidden). Fix this.
- [ ] Height of 256 CSS pixels:
    - [ ] [❗] Some buttons in the multi-note selection menu can't be accessed (clipped).
- [x] [❗] Width of 320 CSS pixels: Some items on the Markdown toolbar can't be accessed -- https://github.com/laurent22/joplin/pull/11636 
- [ ] [🟡] Width of 320 CSS pixels: Some text in certain setting screens can only be read by scrolling both horizontally and vertically.

### 1.4.11: [Non-text contrast](https://www.w3.org/WAI/WCAG22/quickref/#non-text-contrast) (AA)

> The visual presentation of the following have a contrast ratio of at least 3:1 against adjacent color(s):
> 
> - **User Interface Components:** Visual information required to identify user interface components and states, except for inactive components or where the appearance of the component is determined by the user agent and not modified by the author;
> - **Graphical Objects:** Parts of graphics required to understand the content, except when a particular presentation of graphics is essential to the information being conveyed.

- [ ] \[🔎\] Check user interface components (non-text) in the desktop app for contrast in the default theme.

### 1.4.12: [Text spacing](https://www.w3.org/WAI/WCAG22/quickref/#text-spacing) (AA)

Users should be able to override Joplin's styles (through OS accessibility settings?) in the following ways and still use the app:

- Set line spacing to at least 1.5 times the font size.
- Set spacing following paragraphs to at least 2 times the font size.
- Set word spacing to at least 0.16 times the font size.

In particular, [content should not be clipped](https://www.w3.org/WAI/WCAG22/Techniques/failures/F104.html).

- [x] \[🔎\] How can this be done with Joplin (if at all)? Is Joplin still usable in this case?

**Note:** I believe this guideline is mostly relevant for content in web pages? In Joplin the note content can be customised using userstyle.css. UI elements however probably cannot be changed that much without becoming unusable. It still can be customised using chrome.css though.

### 1.4.13: [Content on Hover or Focus](https://www.w3.org/WAI/WCAG22/quickref/#content-on-hover-or-focus) (AA)

When content (e.g. tooltips) can be shown by hovering with the mouse pointer, it should also be possible to show them with a keyboard.

At present, this isn't possible with most `title=` tooltips (at least on Linux). However, the WCAG includes an exception that might be relevant here:

> **Exception**: The visual presentation of the additional content is controlled by the user agent and is not modified by the author.

As an Electron app, however, Joplin **is** the user agent, and so this might not apply.

- [x] [🟡] Currently, most tooltips can only be shown by hovering with a mouse (see [this Chromium bug](https://issues.chromium.org/issues/40570289)). Tooltips should also be shown on keyboard focus.

## [2\. Operable](https://www.w3.org/TR/WCAG22/#operable)

> User interface components and navigation must be operable.

### 2.1.1. Keyboard (A)

> All functionality of the content is operable through a keyboard interface without requiring specific timings for individual keystrokes, except where the underlying function requires input that depends on the path of the user's movement and not just the endpoints. (Level A)

- [x] [❗] ~~Make it possible to~~ Document how to navigate to the Rich Text Editor toolbar with just a keyboard.
    - Currently, focus skips the toolbar when navigating with <kbd>tab</kbd>.
    - **Update**: This is currently possible with <kbd>alt</kbd>-<kbd>F10</kbd> on Linux and Windows. Some Linux windowing systems (e.g. GNOME) handle <kbd>alt</kbd>-<kbd>F10</kbd> by default. If possible, it may make sense to add an alternate keyboard shortcut for this. See the [TinyMCE accessibility documentation](https://www.tiny.cloud/docs/tinymce/latest/tinymce-and-screenreaders/#keyboard-shortcuts).
    - https://github.com/laurent22/joplin/pull/11764
- [x] Settings:
    - [x] [❗] Keyboard shortcuts can't be reassigned without using a mouse (or importing from a file) -- https://github.com/laurent22/joplin/pull/11232
    - [x] [❗] Font picker suggestions can't be accepted -- https://github.com/laurent22/joplin/pull/11763
    - [x] [❗] Can't switch font picker back and forth between monospace-only -- https://github.com/laurent22/joplin/pull/11763
- [x] [❗] Selecting multiple notes and performing actions (e.g. "export") -- https://github.com/laurent22/joplin/pull/10940, https://github.com/laurent22/joplin/pull/11018
- [x] [❗] Re-ordering notes -- https://github.com/laurent22/joplin/pull/10940
- [ ] [❗] Selecting multiple non-adjacent notes and performing actions (e.g. "tags")
    - [ ] Currently, <kbd>shift</kbd>+<kbd>arrow keys</kbd> allows selecting multiple adjacent notes. However, it doesn't seem possible to select non-adjacent notes without first changing the order.
- [x] [❗] Collapse/uncollapse the "NOTEBOOKS" and "TAGS" headers in the sidebar -- https://github.com/laurent22/joplin/pull/11196
- [x] [❗] Make it possible to re-assign keyboard shortcuts without using a mouse. -- https://github.com/laurent22/joplin/pull/11232.
- [x] \[🔎\] Currently, it's difficult to move the keyboard focus to the "new note"/"new to-do"/"sync" buttons
    - [x] <kbd>tab</kbd> currently skips over these controls. However, it is often possible to reach the control through use of <kbd>shift</kbd>\-<kbd>tab</kbd> from the correct element -- https://github.com/laurent22/joplin/pull/11446.
- [x] [❗] Change application layout screen: Make it possible to re-order main screen panels without using a mouse -- https://github.com/laurent22/joplin/pull/11718.
- [ ] [❗] Make it possible to resize app panels without using a mouse.
- [x] [❗] Make it possible to edit code blocks from the Rich Text Editor without using a mouse -- https://github.com/laurent22/joplin/pull/11727
- [x] [❗] Rich Text Editor right-click menus can have incorrect content when opened from a keyboard -- https://github.com/laurent22/joplin/pull/12076

### 2.1.2. No Keyboard Trap (A)

> If keyboard focus can be moved to a component of the page using a keyboard interface, then focus can be moved away from that component using only a keyboard interface, and, if it requires more than unmodified arrow or tab keys or other standard exit methods, the user is advised of the method for moving focus away. (Level A)

- [x] \[🔷\] CodeMirror 6 editor: Move the focus when pressing <kbd>shift</kbd>\-<kbd>tab</kbd> at the beginning of the editor, without need for a leading <kbd>esc</kbd> modifier.
- [x] [❗] Make the CodeMirror 6-based markdown editor the default (https://github.com/laurent22/joplin/pull/10796).
    - [x] The CodeMirror 5 editor traps focus.
- [x] [❗] Rich Text Editor & CodeMirror: Allow escaping focus with keys that don't rely on specific timings (<kbd>ctrl</kbd>\-<kbd>m</kbd> seems to be a standard way of switching focus. See, for example, VSCode.) -- https://github.com/laurent22/joplin/pull/11717.
    - §2.1 includes the "without specific timings" requirement.
    - Currently, <kbd>esc</kbd> then <kbd>tab</kbd> works, but if there's a long delay between the <kbd>esc</kbd> and the <kbd>tab</kbd>, it doesn't work (at least in the Markdown editor). 
- [x] [❗] Rich Text Editor's code block dialog: Make it possible to escape the focus trap (without also closing the dialog) -- https://github.com/laurent22/joplin/pull/11717

### 2.1.3. Keyboard (No Exception) (AAA):

Equivalent to §2.1.1, but removes exceptions.

### 2.1.4. [Character Key Shortcuts](https://www.w3.org/WAI/WCAG22/quickref/#character-key-shortcuts) (A)

If there are unmodified keyboard shortcuts implemented with just a letter/punctuation/special character, then at least one of the following must be true:

- It can be turned off.
- The shortcut can be remapped to some other combination.
- The shortcut is specific to a user interface component and is **active only when it has focus**.

Joplin doesn't seem to have character key shortcuts.

### 2.2.1. [Timing Adjustable](https://www.w3.org/WAI/WCAG22/quickref/#timing-adjustable) (A)
> For each time limit that is set by the content, at least one of the following is true:
> - **Turn off**: The user is allowed to turn off the time limit before encountering it; or
> - **Adjust:** The user is allowed to adjust the time limit before encountering it over a wide range that is at least ten times the length of the default setting; or
> - **Extend:** The user is warned before time expires and given at least 20 seconds to extend the time limit with a simple action (for example, "press the space bar"), and the user is allowed to extend the time limit at least ten times; or
> - **Real-time Exception:** The time limit is a required part of a real-time event (for example, an auction), and no alternative to the time limit is possible; or
> - **Essential Exception:** The time limit is essential and extending it would invalidate the activity; or
> - **20 Hour Exception:** The time limit is longer than 20 hours

- [ ] [❗] "Note deleted" toasts.
- [ ] [❗] Plugin toasts.

### 2.2.2. Pause, Stop, Hide

> For moving, blinking, scrolling, or auto-updating information, all of the following are true: (Level A)

This includes:

> **Auto-updating**: For any auto-updating information that (1) starts automatically and (2) is presented in parallel with other content, there is a mechanism for the user to pause, stop, or hide it or to control the frequency of the update unless the auto-updating is part of an activity where it is essential.

- [ ] [🟡] Provide a way to hide synchronization status while keeping the sidebar open. (Issue https://github.com/laurent22/joplin/issues/8611)
    - [ ] Currently, it **is** possible to hide the sync status button by hiding the sidebar. However, this also involves hiding the note list (so it's not possible to hide sync status without also hiding content that it's "presented in parallel with"). There should be a way to hide the sync status animation without hiding the sidebar.

**§2.2.3-2.2.6** (Level AAA): Concerned with interruptions and timing.

### 2.3. [Seizures: Don't design content in a way that is known to cause seizures](https://www.w3.org/TR/WCAG20/#seizure)

> **§2.3.1** Three Flashes or Below Threshold: Web pages do not contain anything that flashes more than three times in any one second period, or the flash is below the general flash and red flash thresholds. (Level A)

- [x] \[🔎\] Is it possible to cause the UI to flash more than three times? (Check cases involving dark mode transitioning to light mode and switching notes with resources that need to load).
    - [x] Doesn't seem possible 
- §2.3.2 (Level AAA) seems to be equivalent to §2.3.1, but removes the "flash is below the general flash and red flash thresholds" exception.

### 2.4. Navigable

> Provide ways to help users navigate, find content, and determine where they are.

### 2.4.1. [Bypass Blocks](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1&showtechniques=242%2C243%2C241#bypass-blocks) (Level A)

> A mechanism is available to bypass blocks of content that are repeated on multiple Web pages.

Joplin currently has menu items and keyboard shortcuts to jump focus to the note list, sidebar, note title, and note editor (Go > Focus > ...).

### 2.4.2. [Page Titled](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1&showtechniques=243%2C242#page-titled) (Level A)

> Web pages have titles that describe topic or purpose.

- \[🔎\] Research whether Joplin needs to change its title to match the current note.
   - Joplin currently *does* change its title when switching to settings or some other non-note screen.

### 2.4.3. [Focus Order](https://www.w3.org/TR/WCAG22/#focus-order) (Level A)

> If a Web page can be navigated sequentially and the navigation sequences affect meaning or operation, focusable components receive focus in an order that preserves meaning and operability.  
> — https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1&showtechniques=243#focus-order

**Related (§2.4.3)**: https://github.com/laurent22/joplin/pull/10817

- [x] [❗] Don't skip focusable elements on <kbd>tab</kbd>. Instead, use grouping (and/or intelligent use of <kbd>Up</kbd>/<kbd>Down</kbd> arrow keys?)
    - [x] Fix pressing <kbd>Tab</kbd> from the note title input skips the note toolbar -- https://github.com/laurent22/joplin/pull/10932.
    - [x] Fix pressing <kbd>Shift</kbd>\-<kbd>Tab</kbd> from the sidebar skips several focusable items ("add tags", "sync", etc.) -- https://github.com/laurent22/joplin/pull/11641
    - [x] Fix pressing <kbd>Tab</kbd> from the note sidebar skips the note list controls -- https://github.com/laurent22/joplin/pull/11446.
    - [x] Fix pressing <kbd>Tab</kbd> from the note sidebar when multiple notes are selected does nothing -- https://github.com/laurent22/joplin/pull/11018.
    - [x] Fix pressing <kbd>Shift</kbd>\-<kbd>Tab</kbd> from the note list jumps to the sidebar, skipping note list controls -- https://github.com/laurent22/joplin/pull/11446
        - Currently, it's difficult to focus the note list controls using the keyboard.
        - Also reported here: https://github.com/laurent22/joplin/issues/11443
- [ ] Fix focus jumps back to the beginning of the document after using certain controls (see [F85/examples](https://www.w3.org/WAI/WCAG22/Techniques/failures/F85#examples))
    - [ ] [❗] Sort buttons in the note attachments screen.
    - [ ] [❗] Move buttons in the change application layout screen.
- [x] [❗]  #11648
	- [x] Note history screen

### 2.4.4. [Link Purpose (In Context)](https://www.w3.org/WAI/WCAG22/quickref/#link-purpose-in-context) (Level A)

> The purpose of each link can be determined from the link text alone or from the link text together with its programmatically determined link context, except where the purpose of the link would be ambiguous to users in general.  
> — https://www.w3.org/WAI/WCAG22/quickref/#link-purpose-in-context

- Including link information in the same sentence, as is done in this sentence, [is sufficient](https://www.w3.org/WAI/WCAG22/Techniques/general/G53.html).
- \[🔎\] Are there links that don't satisfy these criteria? Are there [links that need an `aria-labelledby` attribute](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA7.html)?
- [x] [❗] Sync wizard: "Click here to select one" link needs additional context -- https://github.com/laurent22/joplin/pull/11649

### 2.4.5. [Multiple Ways](https://www.w3.org/WAI/WCAG22/quickref/#multiple-ways) (Level AA)

> More than one way is available to locate a Web page within a set of Web pages except where the Web Page is the result of, or a step in, a process.

- Most (all?) screens that can be accessed through menus can be accessed through GotoAnything.
- It's possible to access the full settings screen through menus, GotoAnything, and <kbd>ctrl</kbd>-<kbd>,</kbd>.
	- It may not be possible, however, to access individual settings tabs in multiple different ways. Are these tabs considered "part of the settings page"?

### 2.4.6. [Headings and Labels](https://www.w3.org/WAI/WCAG22/quickref/#headings-and-labels) (Level AA)

> Headings and labels describe topic or purpose.

- [x] [🟡] **Headings/main screen**: Joplin currently doesn't use a `<h1>`/etc. tag for the "NOTEBOOKS" heading. This may need to be changed.
    - The sidebar is currently marked with `role=navigation` and has an `aria-label`. As such, marking "NOTEBOOKS" as a heading **might not be necessary** (or even be distracting/redundant). 
- [x] [🟡] **Headings/sync**: Add headings to the various sync authentication pages. See https://github.com/laurent22/joplin/issues/11648.
- [x] [🟡] **Headings/sync**: Improve labels/headings on the sync wizard page. There are multiple buttons labeled "Select" and it's difficult to tell which is associated with which sync target -- https://github.com/laurent22/joplin/pull/11649
- [x] 🟡] Incorrect note viewer label -- https://github.com/laurent22/joplin/pull/11744

### 2.4.7. [Focus Visible](https://www.w3.org/WAI/WCAG22/quickref/#focus-visible) (Level AA)

> Any keyboard operable user interface has a mode of operation where the keyboard focus indicator is visible.

- [x] [🟡] Fix focus not visible on `<select>` elements -- https://github.com/laurent22/joplin/pull/11177
    - See [F78](https://www.w3.org/WAI/WCAG22/Techniques/failures/F78.html).
- [x] [🟡] Fix focus not visible on items in the Rich Text Editor toolbar.
- [ ] [🟡] Style the folder/tag sidebar differently when it has focus. At present, the selected notebook is always indicated, but it isn't clear when the sidebar has focus vs something else.

### 2.4.9. [Link Purpose (Link Only)](https://www.w3.org/WAI/WCAG22/quickref/#link-purpose-link-only) (Level AAA)

> A mechanism is available to allow the purpose of each link to be identified from link text alone, except where the purpose of the link would be ambiguous to users in general.

Based on the "sufficient techniques" section and the "ambiguous to users in general", this seems to be primarily targeted at screen reader users (e.g. when cycling through the links available on a page). To meet this, we might start by:

- [ ] \[🔷\] Adding and visually hiding extended link text in the "Note attachments" screen.
- [ ] \[🔷\] Adding and visually hiding extended link text for the "location" link on the note properties screen.

Also to research:

- [ ] \[🔎\] When navigating with a screen reader, is it possible to tell what the plugin screen header links do?

### 2.4.10. [Section Headings](https://www.w3.org/WAI/WCAG22/quickref/#section-headings) (Level AAA)

> Section headings are used to organize the content.
> 
> Note 1: "Heading" is used in its general sense and includes titles and other ways to add a heading to different types of content.
> 
> Note 2: This success criterion covers sections within writing, not user interface components. User Interface components are covered under Success Criterion 4.1.2.

Given the second note above, this section likely applies mostly to Joplin's welcome notes and settings screen. At present, the welcome notes use headings.

### 2.4.11. [Focus Not Obscured (Minimum)](https://www.w3.org/WAI/WCAG22/quickref/#focus-not-obscured-minimum) (Level AA)

> When a user interface component receives keyboard focus, the component is not entirely hidden due to author-created content.
> 
> - Note 1: Where content in a configurable interface can be repositioned by the user, then only the initial positions of user-movable content are considered for testing and conformance of this Success Criterion.
> - Note 2: Content opened by the user may obscure the component receiving focus. If the user can reveal the focused component without advancing the keyboard focus, the component with focus is not considered hidden due to author-created content.

- [x] [❗] Opening a dialog often obscures focus: https://github.com/laurent22/joplin/pull/10801, and previously https://github.com/laurent22/joplin/pull/10779.
- [x] [❗] Import/export modal obscures focus -- fixed as a part of https://github.com/laurent22/joplin/pull/11181
- [ ] \[🔷\] \[🔎\] Check for other cases where the focus can be hidden by another item. (Trash notification?)

### 2.4.12. [Focus Not Obscured (Enhanced)](https://www.w3.org/WAI/WCAG22/quickref/#focus-not-obscured-enhanced) (Level AAA)

> When a user interface component receives keyboard focus, no part of the component is hidden by author-created content.

2.4.12 is equivalent to 2.4.11, but removes exceptions. See the task list for § 2.4.11.

### 2.4.13. [Focus Appearance](https://www.w3.org/WAI/WCAG22/quickref/#focus-appearance) (Level AAA)

Skipped (level AAA).

### 2.5.1. [Pointer Gestures](https://www.w3.org/WAI/WCAG22/quickref/#pointer-gestures) (Level A)

> All functionality that uses multipoint or path-based gestures for operation can be operated with a single pointer without a path-based gesture, unless a multipoint or path-based gesture is essential.

- Drag and drop for notes -- it seems to be possible to access all related functionality in other ways.
    - Joplin currently also supports reordering notes with alt+arrow keys.
    - Dragging, then dropping, a note into the editor inserts a Markdown link. Right-clicking on the note and selecting "copy Markdown link", then pasting this link into the editor, can achieve the same effect.
    - Notes can be moved to different notebooks by right-clicking, then selecting "move to notebook".

Also:
- [x] [❗] Support moving notebooks without drag and drop (context menu) -- https://github.com/laurent22/joplin/pull/11039
    

### 2.5.2. [Pointer Cancellation](https://www.w3.org/WAI/WCAG22/quickref/#pointer-cancellation) (Level A)

> For functionality that can be operated using a single pointer, at least one of the following is true:
> 
> - **No Down-Event**: The down-event of the pointer is not used to execute any part of the function;
> 
> - **Abort or Undo**: Completion of the function is on the up-event, and a mechanism is available to abort the function before completion or to undo the function after completion;
> - **Up Reversal**: The up-event reverses any outcome of the preceding down-event;
> - **Essential**: Completing the function on the down-event is essential.

- **Note dragging**: Dragging the note back to its original location cancels the action.
- **Notebook dragging**: Dragging a notebook back to its original location (before lifting the pointer) cancels the action.

- [x] [🟡] Clicking on a dialog's dimmed background seems to only be cancellable by dragging the pointer outside the main Joplin window.  
    - Clicking and dragging the dialog's dimmed backdrop into the main dialog content still causes the dialog to close.
    - This is marked as "should" despite being associated with a level A item. The gesture is cancellable by dragging the pointer outside the target element. However, a user may not expect the target element to be the full dialog element.
    - https://github.com/laurent22/joplin/pull/11765
    

> [!NOTE]
> 
> A search for `mousedown` and `pointerdown` in Joplin's GitHub repository results in very few matches:
> 
> - [Forwarding a `mousedown` event](https://github.com/laurent22/joplin/blob/d1ac3d415ea7665271accd581807e1e5ae56feef/packages/editor/CodeMirror/CodeMirror5Emulation/CodeMirror5Emulation.ts#L99) for the editor's plugin compatibility layer.
> - Matches in the web clipper.
> - No matches for `pointerdown`
> 
> This suggests (but does not imply) that Joplin desktop doesn't directly handle pointer down events.

### 2.5.3. [Label in Name](https://www.w3.org/WAI/WCAG22/quickref/#label-in-name) (Level A)

> For user interface components with labels that include text or images of text, the name contains the text that is presented visually.

- \[🔎\] Check for [F96](https://www.w3.org/WAI/WCAG22/Techniques/failures/F96.html) in all controls with a visible label (verify that no visible text is excluded from the label, perhaps using a screen reader). When labeling controls, be sure not to cause issues related to this requirement.

**Welcome notes**:
- [x] [❗] https://github.com/laurent22/joplin/pull/11643

### 2.5.4. [Motion Actuation](https://www.w3.org/WAI/WCAG22/quickref/#motion-actuation) (Level A)

> Functionality that can be operated by device motion or user motion can also be operated by user interface components and responding to the motion can be disabled to prevent accidental actuation, except when:
> 
> - Supported Interface: The motion is used to operate functionality through an accessibility supported interface;
> - Essential: The motion is essential for the function and doing so would invalidate the activity.

Joplin for desktop seems to not have functionality that can be activated by device motion (e.g. accelerometer). 

### 2.5.5. [Target Size (Enhanced)](https://www.w3.org/WAI/WCAG22/quickref/#target-size-enhanced) (Level AAA)

Skipped (Level AAA).

### 2.5.6. [Concurrent Input Mechanisms](https://www.w3.org/WAI/WCAG22/quickref/#concurrent-input-mechanisms) (Level AAA)

Skipped (Level AAA).

### 2.5.7. [Dragging Movements](https://www.w3.org/WAI/WCAG22/quickref/#dragging-movements) (Level AA)

> All functionality that uses a dragging movement for operation can be achieved by a single pointer without dragging, unless dragging is essential or the functionality is determined by the user agent and not modified by the author.

Based on the description of §2.5.1, satisfying 2.5.1 seems to imply that 2.5.7 is also satisfied.

Note that the extended description of this section states:

> This requirement is separate from keyboard accessibility because people using a touch screen device may not use a physical keyboard.

...and later...

> Success Criteria 2.1.1 Keyboard and 2.1.3 Keyboard (No Exception) require dragging features to be keyboard accessible. However, achieving keyboard equivalence for a dragging operation does not automatically meet this Success Criterion. It is possible to create an interface that works with dragging and keyboard controls that does not work using only clicks or taps.

For now, see the checklist for §2.5.1.

### 2.5.8. [Target Size (Minimum)](https://www.w3.org/WAI/WCAG22/quickref/#target-size-minimum) (Level AA)

> The size of the target for pointer inputs is at least 24 by 24 CSS pixels, except where:
> 
> - **Spacing:** Undersized targets (those less than 24 by 24 CSS pixels) are positioned so that if a 24 CSS pixel diameter circle is centered on the bounding box of each, the circles do not intersect another target or the circle for another undersized target;
> - **Equivalent:** The function can be achieved through a different control on the same page that meets this criterion;
> - **Inline:** The target is in a sentence or its size is otherwise constrained by the line-height of non-target text;
> - **User agent control:** The size of the target is determined by the user agent and is not modified by the author;
> - **Essential:** A particular presentation of the target is essential or is legally required for the information being conveyed.

- Markdown toolbar buttons are 26 by 26 CSS pixels.

To change:
- [ ] Reverse sort order is 18 px x 26 px
    - This function can also be accessed from view > sort notes by > reverse sort order.
- [ ] Settings checkboxes are 14 px x 14 px. However, 1) checkbox labels are larger and can also be clicked, 2) the "spacing" requirement should be satisfied due to spacing between setting controls.
- [ ] [🟡] Note list item checkboxes are 14px x 14px and are contained in a clickable container. Provide an alternate way to toggle to-dos or increase the checkbox size.
- [x] [🟡] <span id="scrollbar-width">Scrollbars</span> have 7px width ([issue report](https://discourse.joplinapp.org/t/sliders/42039/2)).
	- This is less than the minimum dimension of 24px. See https://developer.mozilla.org/en-US/docs/Web/CSS/::-webkit-scrollbar#accessibility.
	- The WCAG lists "Providing a mechanism to change the size of the target independent of magnification" as a sufficient technique.
	- See https://github.com/laurent22/joplin/pull/11627

### 3.1.1. [Language of Page](https://www.w3.org/WAI/WCAG22/quickref/#language-of-page) (Level A)

> The default human language of each Web page can be programmatically determined.

For HTML, [H57](https://www.w3.org/WAI/WCAG22/Techniques/html/H57.html) recommends using the `lang` attribute. Among other reasons, this is so that

> - Speech synthesizers that support multiple languages will be able to orient and adapt to the pronunciation and syntax that are specific to the language of the page, speaking the text in the appropriate accent with proper pronunciation.

- [x] [❗] Set the HTML `lang` attribute to match Joplin's current locale -- https://github.com/laurent22/joplin/pull/11218.

### 3.1.2. [Language of Parts](https://www.w3.org/WAI/WCAG22/quickref/#language-of-parts) (Level AA)

> The human language of each passage or phrase in the content can be programmatically determined except for proper names, technical terms, words of indeterminate language, and words or phrases that have become part of the vernacular of the immediately surrounding text.

- [ ] [🟡] Some (or most) of Joplin's supported UI languages aren't complete translations and fall back to the default English locale in places. Wrap these places with `lang="en"`.
    - I'm not sure how feasible this is — many localized strings are included in places that can't easily be marked as a different language from the surrounding text (not easy to set HTML props on the container based on failed/successful localization).

### 3.1.3. [Unusual words](https://www.w3.org/WAI/WCAG22/quickref/#unusual-words) (Level AAA)
> A mechanism is available for identifying specific definitions of words or phrases used in an unusual or restricted way, including idioms and jargon.

- [ ] [🔷] Joplin's UI uses technical terms in some locations, particularly in settings. It may make sense to link these terms to sections of Joplin's documentation. Examples of terms that could be explained include:
	- "End-to-end encryption" in settings.
	- "Master password".
	- "Disabled keys".
	- "Stylesheet".
	- "Markdown renderer".
	- Names of Markdown plugins.
	- "Proxy"

### 3.1.4. [Abbreviations](https://www.w3.org/WAI/WCAG22/quickref/#abbreviations) (Level AAA)

Skipped (Level AAA)

### 3.1.5. [Reading Level](https://www.w3.org/WAI/WCAG22/quickref/#reading-level) (Level AAA)

Skipped (Level AAA)

### 3.1.6. [Pronunciation](https://www.w3.org/WAI/WCAG22/quickref/#pronunciation)
> A mechanism is available for identifying specific pronunciation of words where meaning of the words, in context, is ambiguous without knowing the pronunciation.

- [ ] [🔷] [🔎] Evaluate Joplin's UI text and determine whether any words have pronunciations that cannot be inferred from context, where different pronunciations would have different meanings (see ["Understanding SC 3.1.6" for details](https://www.w3.org/WAI/WCAG22/Understanding/pronunciation.html)).

### 3.2 - Predictable
> Make Web pages appear and operate in predictable ways.

### 3.2.1. [On Focus](https://www.w3.org/WAI/WCAG22/quickref/#on-focus) (Level A)
> When any user interface component receives focus, it does not initiate a change of context.

> [!NOTE]
>
> The WCAG defines "changes of context" to be:
> > major changes that, if made without user awareness, can disorient users who are not able to view the entire page simultaneously
> 
> It includes "focus" and "content that changes the meaning of the web page" as examples.

- [x] [🔎] Check for focus causing changes in context.

### 3.2.2. [On Input](https://www.w3.org/TR/WCAG22/#on-input) (Level A)
> Changing the setting of any user interface component does not automatically cause a change of context unless the user has been advised of the behavior before using the component.

- [ ] [❗] Using the sort order buttons in the note attachments screen cause a change in context.
- [ ] [🔎] Does pressing buttons in the Markdown editor toolbar (which change editor content) count as creating a [change in context](https://www.w3.org/WAI/WCAG22/Understanding/on-input.html)? If so, [this (closed) pull request](https://github.com/laurent22/joplin/pull/11441) may be related to this success criterion.

### 3.3.1. [Error identification](https://www.w3.org/TR/WCAG22/#error-identification) (A)

> If an [input error](https://www.w3.org/TR/WCAG22/#dfn-input-error) is automatically detected, the item that is in error is identified and the error is described to the user in text.

- [ ] [🔎] The keyboard shortcuts screen shows an error message only when the user hovers over or clicks on the "warning" icon. Is this sufficient?

### 3.3.2. [Labels or instructions](https://www.w3.org/TR/WCAG22/#labels-or-instructions) (A)

> [Labels](https://www.w3.org/TR/WCAG22/#dfn-labels) or instructions are provided when content requires user input.

The description of this success criterion includes:
> In the case of radio buttons, checkboxes, comboboxes, or similar controls that provide users with options, each option must have an appropriate label so that users know what they are actually selecting. 

As such, this may also apply to repeated buttons or other controls used to select options.

- [x] [❗] Sync wizard: "Select" buttons are repeated and unlabelled -- https://github.com/laurent22/joplin/pull/11649
- [x] [❗] Note attachments screen: Delete buttons all have the same label.
   - When using a screen reader, it's currently difficult to tell which delete button deletes which attachment.
   - https://github.com/laurent22/joplin/pull/11749

### 3.3.3. [Error Suggestion](https://www.w3.org/WAI/WCAG22/quickref/#error-suggestion) (Level AA)
> If an input error is automatically detected and suggestions for correction are known, then the suggestions are provided to the user, unless it would jeopardize the security or purpose of the content.

- [x] [🟡] Display an error (and allow users to correct it) if an invalid date is entered in the note properties dialog. (Fixed by https://github.com/laurent22/joplin/pull/11725 since it is no longer possible to input a invalid date)
- [x] [🟡] Display a warning if the user enters an invalid location in the note properties dialog. (Fixed by https://github.com/laurent22/joplin/pull/11784)
- [ ] [🟡] Display a warning in the color picker dialog in the Rich Text Editor when invalid color components are entered.
	- Currently, the invalid input is highlighted in red and marked with `aria-invalid="true"`. This *might* be sufficient.
- [ ] [🟡] Master password dialog: Inform the user when the password and the password repeat don't match.
	- Currently, the "submit" button is disabled, but there's no explanation for why.
- [ ] [🟡] Built-in backup plugin: Inform the user when password and "confirm password" don't match. (Requires upstream contributions and possibly plugin API changes).

### 3.3.4. [Error Prevention (Legal, Financial, Data)](https://www.w3.org/WAI/WCAG22/quickref/#error-prevention-legal-financial-data) (Level AA)
> For Web pages that cause legal commitments or financial transactions for the user to occur, that modify or delete user-controllable data in data storage systems, or that submit user test responses, at least one of the following is true:
> - **Reversible**: Submissions are reversible.
> - **Checked**: Data entered by the user is checked for input errors and the user is provided an opportunity to correct them.
> - **Confirmed**: A mechanism is available for reviewing, confirming, and correcting information before finalizing the submission.

This guideline is described as applying to systems that "modify or delete user-controllable data in data storage systems". Joplin manages user data. Checking whether Joplin complies with this guideline might involve:
- [ ] [🔎] Are there any actions that aren't reversible (or don't provide users an opportunity to correct/confirm)?

### 3.3.5. [Help](https://www.w3.org/WAI/WCAG22/quickref/#help) (Level AAA)
> Context-sensitive help is available.

This success criterion suggests, for example,
- [Providing a help link on each webpage.](https://www.w3.org/WAI/WCAG22/Techniques/general/G71)
	- Joplin currently provides a help link in its "help" menu, though this may be difficult to discover.
- [Providing a help assistant on the website](https://www.w3.org/WAI/WCAG22/Techniques/general/G193)
- [Spell checking in text fields](https://www.w3.org/WAI/WCAG22/Techniques/general/G194)
- [Including instructions at the beginning of a form](https://www.w3.org/WAI/WCAG22/Techniques/general/G184)

### 3.3.6. [Error Prevention (All)](https://www.w3.org/WAI/WCAG22/quickref/#error-prevention-all) (Level AAA)
> For Web pages that require the user to submit information, at least one of the following is true:
> - **Reversible**: Submissions are reversible.
> - **Checked**: Data entered by the user is checked for input errors and the user is provided an opportunity to correct them.
> - **Confirmed**: A mechanism is available for reviewing, confirming, and correcting information before finalizing the submission.

This is equivalent to SC 3.3.4, but applies to a larger set of applications.

### 3.3.7. [Redundant Entry](https://www.w3.org/WAI/WCAG22/quickref/#redundant-entry) (Level A)
> Information previously entered by or provided to the user that is required to be entered again in the same process is either:
> - auto-populated, or
> - available for the user to select.
>
> Except when:
> - re-entering the information is essential,
> - the information is required to ensure the security of the content, or
> - previously entered information is no longer valid.

The success criterion's extended description also states:
> This Success Criterion does not add a requirement to store information between sessions. A [process](https://www.w3.org/WAI/WCAG22/Understanding/redundant-entry.html#dfn-process) is defined on the basis of an activity and is not applicable when a user returns after closing a session or navigating away. However, a process can run across different domains, so if a check-out process includes a 3rd party payment provider, that would be in scope.

I'm unaware of cases where Joplin requests data re-entry (except in the case of passwords). The success criterion does include search as an example, however. As such,
- [ ] [❗] Pre-populate the Rich Text editor search dialog with the global search content.
	- Search-entry pre-population is currently only done in the Markdown editor.

### 3.3.8. [Accessible Authentication (Minimum)](https://www.w3.org/WAI/WCAG22/quickref/#accessible-authentication-minimum) (Level AA)
> A cognitive function test (such as remembering a password or solving a puzzle) is not required for any step in an authentication process unless that step provides at least one of the following: 
> - **Alternative**: Another authentication method that does not rely on a cognitive function test.
> - **Mechanism**: A mechanism is available to assist the user in completing the cognitive function test.
> - **Object Recognition:** The cognitive function test is to recognize objects.
> - **Personal Content:** The cognitive function test is to identify non-text content the user provided to the Web site.

The guidelines also state that supporting password manager input and allowing copy/paste to reduce re-typing satisfy this criterion.
- [x] Joplin's password inputs should all support pasting from other locations. However, I'm unsure whether direct entry using a password manager is supported.

### 3.3.9. [Accessible Authentication (Enhanced)](https://www.w3.org/WAI/WCAG22/quickref/#accessible-authentication-enhanced) (Level AAA)

The WCAG states:

> This criterion is the same as [Accessible Authentication (Minimum)](https://www.w3.org/WAI/WCAG22/Understanding/accessible-authentication-minimum) but without the exceptions for objects and user-provided content.

Joplin should allow pasting passwords from password managers or other applications when signing in to sync targets that authenticate through the app. Sync targets that authenticate through a web browser (e.g. Joplin Cloud) should also allow using the user's password manager.

## §4 - Robust

> Content must be robust enough that it can be interpreted by a wide variety of user agents, including assistive technologies.

### 4.1.2. [Name, Role, Value](https://www.w3.org/WAI/WCAG22/quickref/#name-role-value) (Level A)

> For all user interface components (including but not limited to: form elements, links and components generated by scripts), the name and role can be programmatically determined; states, properties, and values that can be set by the user can be programmatically set; and notification of changes to these items is available to user agents, including assistive technologies.
>
> *Note 1*: This success criterion is primarily for Web authors who develop or script their own user interface components. For example, standard HTML controls already meet this success criterion when used according to specification.

At present, most ARIA-related tasks are listed under SC 1.3.1. In addition to those, issues include:

- [x] [❗] Correctly convey the structure of the font picker control in settings -- https://github.com/laurent22/joplin/pull/11763
- [x] [❗] Correctly convey the structure of the note list (e.g. with semantic HTML or `role=`).
    - Partial: https://github.com/laurent22/joplin/pull/10940
- [x] [❗] Correctly convey the structure of the notebook and tag list (use [the treeview pattern](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/)).
    - https://github.com/laurent22/joplin/pull/11196
- [x] [❗] Correctly convey the structure of Goto Anything search results.
    - Related: [WAI ARIA Combobox Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/)
    - https://github.com/laurent22/joplin/pull/10816
- [x] [❗] Correctly convey the structure of the settings sidebar.
    - https://github.com/laurent22/joplin/pull/10812 adds additional `aria` information (e.g. `aria-controls`).
- [x] [❗]  Note and Markdown toolbars -- https://github.com/laurent22/joplin/pull/10819
- [x] [❗]  Date/time picker for to-dos -- https://github.com/laurent22/joplin/pull/11725

## 4.1.3. [Status Messages](https://www.w3.org/WAI/WCAG22/quickref/#status-messages) (Level AA)
> In content implemented using markup languages, status messages can be programmatically determined through role or properties such that they can be presented to the user by assistive technologies without receiving focus.

Several status messages need appropriate accessibility information to be exposed:
- [x] [🟡] #11741 -- https://github.com/laurent22/joplin/pull/11776
    - Currently, it's **very** confusing when the note list disappears (with no announcement).
    - One possible solution [is to add a new status indicator](https://github.com/laurent22/joplin/compare/dev...personalizedrefrigerator:pr/desktop/accessibility/announce-visiblepanes-changes). However, with the current status indicator logic, this adds a new focusable item to the end of the document. 
- [x] [🟡] #11742
- [ ] [🟡] Announce changes made by the "toggle editor layout" command — hiding the note viewer when not focused currently isn't announced.
- [ ] [🟡] Sync status message.
	- Marking the sync status region with `role=status` could be distracting when navigating with a screen reader. Implement https://github.com/laurent22/joplin/issues/8611 first.
	- Using `role=log` is one possible alternative.
- [x] [🟡] Searching: Announce changes to results -- https://github.com/laurent22/joplin/pull/11634
- [ ] [🟡] Error messages in settings.
- [x] [🟡]  Importing/exporting modal.
- [ ] [🟡]  Top-of-screen error banner (for sync & decryption errors).
- [ ] [🟡] Changes to the Markdown editor from the toolbar -- https://github.com/laurent22/joplin/pull/11441
- [x] [🟡] Moving panels in the "change application layout" dialog.
    - When it's easier to move the keyboard focus to the "move" buttons (and the buttons are labelled, see https://github.com/laurent22/joplin/pull/11718), this issue will be more noticeable. 
- [ ] [🟡] [🔎] Rich Text Editor -- when selection moves to a code block. (To check: Is this a status update issue or a missing label issue?)
    -  In the Rich Text Editor, when the selection moves to a code block, the Orca screen reader reads nothing (despite Orca reading the code block when it is **not** selected, but on the same line as the cursor). Instead, nothing is read (not even a "selection updated" notification).

**Comment by rkingett** [comment]
I am loving this! I am not sure what the desktop app is programmed in, but I am worried there is an over reliance on ARIA where native controls and functions could be used, such as, the below for example regarding links.

I genuinely do not think links need an ARIALABEL if link text/titles are descriptive enough.

But I am beyond excited to see this progress so will watch closely! If I knew what the mobile apps and desktop apps were made in, I could better advise but this is a great start!

**Comment by tomasz1986** [comment]
> I am worried there is an over reliance on ARIA where native controls and functions could be used, such as, the below for example regarding links.

The general rule is that if you have to use things like `role=img`, then this likely means that the element itself should just be a simple `img` (and not `div` or similar). You can find this ARIA overuse on many websites with `<nav role="navigation>`, which is completely useless and only adds bloat.

**Comment by personalizedrefrigerator** [comment]
> I genuinely do not think links need an ARIALABEL if link text/titles are descriptive enough.

Thank you for the feedback!

There are cases where links/buttons in Joplin have no text content but have a `title` attribute. According [to `w3.org`](https://www.w3.org/WAI/tutorials/forms/labels/#using-the-title-attribute),
> The title attribute can also be used to identify form controls. This approach is generally less reliable and not recommended because some screen readers and assistive technologies do not interpret the title attribute as a replacement for the label element, possibly because the title attribute is often used to provide non-essential information.

Based on this, `aria-label` is being added in some cases where a button/link has no text content, even if it already has a `title`.

**Related links**:
- [Advisory technique: Describing a link's purpose with aria-label](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA8.html)
- [How different screen readers interpret link + title and link + aria-label](https://www.deque.com/blog/text-links-practices-screen-readers/)


> If I knew what the mobile apps and desktop apps were made in, I could better advise but this is a great start!

The desktop app is built with Electron and React. The mobile app is built with React Native.

**Comment by personalizedrefrigerator** [comment]
> The general rule is that if you have to use things like role=img, then this likely means that the element itself should just be a simple img (and not div or similar).

Thank you for the feedback!

Currently, Joplin **does** use `role="img"` for its font icons. For example, Joplin uses FontAwesome `<i>` webfont icons. For this type of icon, [ARIA 24](https://www.w3.org/WAI/WCAG22/Techniques/aria/ARIA24.html) recommends:
> Some low vision users need user stylesheets, scripts, or extensions to override the fonts on a page to perceive text content. However, they need to be able to perceive icon fonts that are meaningful, such as a star signifying a favorite, or an email icon in a link.
>
> The key is for the author to semantically markup icon fonts with role="img". Then the user's font replacement selector can hook into that semantic and exclude role="img". This results in those icon fonts being shown.

An alternative to webfont+`role=img` could be to use SVG icons (e.g. with [FontAwesome SVG](https://docs.fontawesome.com/web/dig-deeper/webfont-vs-svg)). However, this would likely require significant code changes.

**Comment by tomasz1986** [comment]
Yeah, icon fonts are not good for accessibility since they are basically a hack to display images instead of text (as should be expected with fonts normally), so they have issues with screen readers, and they also cause problems when the user wants (or needs due to a particular disability) to use their own font for the whole application or website. SVG icons are much better in this aspect, but as you've said, implementing them would require a major rewrite of the code 🙁.

**Comment by laurent22** [comment]
> [🔎] How can we make it easy for users to add alternative text to videos and audio elements?

**[Suggest close]** When attaching a video, it puts the filename as the title, eg `[myvideo.mp4](:/5ca9d953c01643d6a503b473c1e92a2d)`. Then in the renderer this is used for the link `title` attribute and for the content: `<a href="..." title="myvideo.mp4">myvideo.mp4</a>`. Isn't it sufficient? Beyond this, I believe it is not the app responsability to make the user content accessible - they have the option to give detailed description in the title or not, but we can't control that, so I don't think there is more to do? Or maybe putting the text in the `alt` attribute too?

> [🔎] Does Joplin use ::before or ::after to include non-decorative content (see F87)?

We use `::before` to render the checkboxes in the RTE checkbox lists. I suspect it would be difficult to improve this.

> [🔎] Does Joplin use structural markup to achieve a visual-only effect? (F43 and related).

**[Suggest close]** Very hard to say but generally we use semantic markup so I would think there are not many such instances. The link provides example of using `blockquote` for indentation or `h4` to make something bold and we generally don't do this. I think it's safe to tick that item off now.

> [🔎] Meaningful sequences can be particularly easy to disrupt using flex-order or similar order-changing CSS. Are we using flex-order anywhere? See F1.

**[Suggest close]** Same comment as above - as the HTML is relatively clean I suspect we don't have too many strange out of order sequences. We don't use `flex-order` as far as I can see.

> [🔎] Does this include RTL support? (Related issue).

**[Suggest close]** We do not officially support RTL for now. If certain things work it's great but I know there are still issues and fixing this would be a project of its own.

> (Different task) Check the documentation website.

**[Suggest close]** I've searched for basic colours in the documentation - green, blue, etc. and didn't find any instructions based on colours. Only possible exception is [here](https://joplinapp.org/help/api/tutorials/cm6_plugin/#style-the-active-line) where it says "The active line should now have a light-yellow background". But I'm not sure how that could be improved. Possibly also out of scope for this project (which is about accessibility for the end user rather than for developers).

> [🔎] Does this require better by-default support for small screens? If so, a possible solution could be to minimize secondary panels (e.g. by moving them to an overflow tab) when low on space.

**[Suggest close]** I'd think this is out of scope for the desktop app because we provide the mobile app for smaller devices

> [🔎] Check color contrast between neighboring content.

> [🔎] Check user interface components (non-text) in the desktop app for contrast in the default theme.

> Add a high-contrast theme for the markdown editor?

I think these three items are similar? Basically about having sufficient contrasts between adjacent elements.

I couldn't actually find a tool to check this. The [WebAIM contrast checker](https://webaim.org/resources/contrastchecker/) is to compare foreground and background colour, which is not the same as comparing, for example, the colour the selected notebook background against the whole sidebar background colour. We need something to test adjacent colours it seems.

We have a visible focus indicator in any case which would be part of the Sufficient Techniques: "G195: Using an author-supplied, visible focus indicator"

> 1.4.12: Text spacing: [🔎] How can this be done with Joplin (if at all)? Is Joplin still usable in this case?

I think that's for web pages where user can provide a user CSS stylesheet? 

As a test I think we should try to set a custom stylesheet for the Markdown editor, RTE and viewer with the mentioned properties and see if it still works. If it doesn't we should evaluate if there are simple fixes we can implement - if it's too complex we should drop it.

And we shouldn't support this for the rest of the interface. The guideline mentions "content" which I think is understood to be articles, etc., not user interface elements.

> [🔎] Currently, it's difficult to move the keyboard focus to the "new note"/"new to-do"/"sync" buttons

**[Suggest close]** Not sure if something has changed but I was able to move the focus to all these buttons with both TAB and Shift+TAB

> [🔎] Is it possible to cause the UI to flash more than three times? (Check cases involving dark mode transitioning to light mode and switching notes with resources that need to load).

**[Suggest close]** I tried and couldn't get anything to flash three times. Including loading HTML notes. So I'd say we can skip this one.

> [🔷] [🔎] Check for other cases where the focus can be hidden by another item. (Trash notification?)

**[Suggest close]** I'm not sure I understand this guideline. But using the keyboard I could give focus to the notification "Cancel" button as would be expected. That notification also doesn't hide enough of the content or for long enough to be a problem.

> 2.4.13. Focus Appearance (Level AAA)
>
> 2.5.5. Target Size (Enhanced) (Level AAA)
>
> 2.5.6. Concurrent Input Mechanisms (Level AAA)

Should we skip the AAA items?

> [🔎] Check for focus causing changes in context.

**[Suggest close]** Even without accessibility in mind that would be bad design probably, and I can't think of any such instance.

> [🔎] The keyboard shortcuts screen shows an error message only when the user hovers over or clicks on the "warning" icon. Is this sufficient?

I'm not sure why it was done this way, and I don't feel it's very good. We should change it to display the message below the keyboard shortcut, unless there's a good reason for the current UI.

**Comment by zersiax** [comment]
Am a screen reader user and developer, would it be helpful at this point to give the desktop app a quick runthrough to find blockers?

**Comment by personalizedrefrigerator** [comment]
> Am a screen reader user and developer, would it be helpful at this point to give the desktop app a quick runthrough to find blockers?

That would be very helpful! 

There have been a few accessibility-related changes since the last v3.2 release. If you test on v3.2.12, here are some of the unreleased screen-reader-related changes to be aware of:
<details><summary>Desktop: Recent screen-reader-related changes</summary>

- **Welcome notes**: [Improved descriptions in welcome notes](https://github.com/laurent22/joplin/commit/d0b783c595cdcd272ead8eb7819862e2f89b3cd4)
- **Search**: [Improvement in how search results are announced](https://github.com/laurent22/joplin/commit/157ad2c0cd16a586312b8140e66592b3b270577a).
- **Tab navigation**: [Fixed: Pressing Shift+Tab when focus is on notebook list would jump straight to editor](https://github.com/laurent22/joplin/commit/29e55b8231df2769f5e4ace4da0efe78cd7ee8d4)
- **Search**: [Marked a search input as a search input](https://github.com/laurent22/joplin/commit/dc10ff6215930ec2a86622e3d9c2c540ce7b078f)
- **Sync wizard**: [Fixed unlabeled images, buttons with the same description in the sync wizard](https://github.com/laurent22/joplin/commit/e8e3ef36ed4226b4cd683c294ad04079126463b8)
- **Okay/cancel/accept buttons**: [Marked navigation bars in secondary screens (settings, note properties) as navigation regions](https://github.com/laurent22/joplin/commit/e1b41cff5f110d798385b3099dbe5f252c1c787c).
- **Toolbar**: [Labelled an unlabelled toolbar button in the Rich Text Editor](https://github.com/laurent22/joplin/commit/c357b77a48462e01490cd7824a932e21c288c2e8).

</details>

**Edited**: Changed v3.2.11 to v3.2.12 as the stable release version. This version is built from the `release-3.2` branch and doesn't have the accessibility fixes listed in the "recent changes" dropdown.

**Comment by personalizedrefrigerator** [comment]
I'm linking [screen reader accessibility feedback provided on the Joplin Forum](https://discourse.joplinapp.org/t/calling-visually-impaired-users-help-us-improve-accessibility/42094/5?u=personalizedrefriger).

**Comment by zersiax** [comment]
@personalizedrefrigerator  would you be able to toss me a dist of the windows build of that branch? I'd rather test with latest, but I spent a few hours trying to get it to build and was not really getting anywhere with it :) if that's tricky I can test with stable, just figured it'd be aeasier to work with something that already has the latest fixes integrated so I'm not covering old ground as it were

**Comment by personalizedrefrigerator** [comment]
> @personalizedrefrigerator would you be able to toss me a dist of the windows build of that branch? I'd rather test with latest, but I spent a few hours trying to get it to build and was not really getting anywhere with it :) if that's tricky I can test with stable, just figured it'd be aeasier to work with something that already has the latest fixes integrated so I'm not covering old ground as it were

I'll look into creating a new 3.3 pre-release after some of the recent accessibility pull requests have been merged! 

**Update**: This might take a day or two — there are two major regressions that should be fixed first ([a crash](https://github.com/laurent22/joplin/pull/11737) and [missing menu items](https://github.com/laurent22/joplin/issues/11738)). **Update 2**: Waiting for #11737 to be reviewed...

**Comment by zersiax** [comment]
I've been playing with the latest release and so far, very much liking what I'm seeing. There is one question I have that I might just be misunderstanding.
The list of notebooks is a treeview, in that your notebooks can contain notebooks. I'm coming from CherryTree, where the notes within the notebooks also show up as treeview items within the given notebook, so a right arrow drills down into the list of notes for the given notebook.
In Joplin, I'm noticing that you're meant to rpess Enter on a notebook, t which point you get routed to the list of notes for that notebook. This works well, but I didn't immediately see how you're meant to quickly get back to the list of notebooks from there, this takes several shift+tabs or tabs.
IDEA: make right arrow on a notebook that has no sub-notebooks move to the note list, and left arrow in the note list move back to the notebook list, given those keys don't currently do anything in those views anyway. If there's already an established way to do this I missed it, feel free to set me straight :)

**Comment by personalizedrefrigerator** [comment]
> If there's already an established way to do this I missed it, feel free to set me straight :)

Here are two ways:
- <kbd>shift</kbd>-<kbd>enter</kbd>
- <kbd>ctrl</kbd>-<kbd>shift</kbd>-<kbd>s</kbd> (possibly different on Mac).
    - See the "Focus" submenu of "Go" in the app menubar.

I think @pedr has recommended creating a walkthrough or tutorial for the keyboard shortcuts available in the app.

> IDEA: make right arrow on a notebook that has no sub-notebooks move to the note list, and left arrow in the note list move back to the notebook list, given those keys don't currently do anything in those views anyway.

Thank you for the suggestion!

<details><summary>Related notes</summary>

- Left/right arrow keys is probably much more discoverable than the current keyboard shortcuts.
- It's possible to move the notes list above, below, and to the left of the notebook sidebar. (Though there are significant accessibility issues with the UI for this).

</details>

**Comment by personalizedrefrigerator** [comment]
> @personalizedrefrigerator would you be able to toss me a dist of the windows build of that branch? I'd rather test with latest, but I spent a few hours trying to get it to build and was not really getting anywhere with it :)

**Update**: Making an official prerelease is taking a while (related to [this issue](https://github.com/laurent22/joplin/pull/11737)). If you would like help building the Windows version, I would be happy to assist.

**Comment by kaushalv17** [comment]
I'd like to work on the Pre-populate the Rich Text editor search dialog with the global search content item (SC 3.3.7). Planning to submit a PR for this as part of my GSoC 2026 application. Is this still available to work on?

**Comment by kaushalv17** [comment]
Update: PR submitted — https://github.com/laurent22/joplin/pull/14936

**Comment by kaushalv17** [comment]
Hi, I'd like to work on adding alt text editing support for images in the Rich Text Editor.
From the checklist above, I can see that SC 1.1.1 (Non-text content) is still incomplete because there's currently no way to set/edit alt text for images in the RTE. My plan:

Add an "Edit image description (alt text)" option to the right-click context menu when clicking on an image in the RTE
Show a simple dialog where the user can type/update the alt text
Persist it back to the underlying Markdown as ![alt text](:/resourceId)

Let me know if this approach makes sense or if there's a preferred implementation direction. Happy to discuss before opening a PR.

**Comment by zersiax** [comment]
Just cautioning that clicking is not something everybody is able to do. I'm relatively certain just highlighting the spot where the image appears within the RTE using the keyboard is going to make this option appear in the context menu, but make sure to test that :) @kaushalv17

**Comment by kaushalv17** [comment]
Thanks for the heads up @zersiax! Will make sure to test keyboard navigation as well before opening the PR 👍
